### PR TITLE
[CR] Seamless loading of gzipped world/maps/x.y.z/x.y.z.map files

### PR DESCRIPTION
**Pull request by int-ua**
#### Superseded
by #44218

#### Summary

SUMMARY: Performance "Seamless loading of gzipped world/maps/x.y.z/x.y.z.map files"

#### Purpose of change

Ability to drastically reduce save folder sizes with (hopefully) only a marginal performance impact.

#### Describe the solution

Checking first two bytes of files before loading JSON and if they are compressed use `gzstream` instead of `std::ifstream` to read them.
Reference: https://github.com/kevingranade/Cataclysm/commit/e751f0bc4acd52cbbf6e77dc2610d2aa98b232ef

#### Testing

```sh
#!/bin/bash
WORLD_NAME="test"
./cataclysm-tiles --world "$WORLD_NAME"
cp -r "save/${WORLD_NAME}" "save/${WORLD_NAME}_gz"
find "save/${WORLD_NAME}_gz/maps/" -type f -exec gzip "{}" \; -exec mv "{}.gz" "{}" \;
./cataclysm-tiles --world "${WORLD_NAME}_gz"
du -sh "save/${WORLD_NAME}/maps/"
du -sh "save/${WORLD_NAME}_gz/maps/"
```

#### Additional context

#10784 #14307 #19192 #24916 #32530

### CODE REVIEW

**Comment by int-ua** [comment]
Not sure I will have time in the coming days to work more on this PR unfortunately. But I definitely want this feature and no one else seems interested.

**Comment by int-ua** [comment]
`libgzstream-dev` package is not available for Ubuntu Xenial. Is there anything that can be done about that?

**Comment by int-ua** [comment]
Is it ok if I add `gzstream` as a git submodule? It's easier to just go with a folder I guess but would you (maintainers) mind if try submodule?

**Comment by jbytheway** [comment]
Of course it is up to @kevingranade but I'd strongly recommend against using submodules.  In my experience they always cause more issues than they solve.  For example, new people checking out the codebase have extra steps they need to take or things behave strangely.

You already ran into a related issue in needing to alter the actions to checkout the submodules.  There are so many little places where similar issues will arise; it's just not worth the trouble.

**Comment by int-ua** [comment]
In my opinion the issue is highly outweighed by the benefits of not forcing everyone to download the history of everything. Although it's quite late at this point I still think that this project could benefit from git submodules and I'm trying to establish a precedent. It's just an additional command or two to run that can save you g~~h~~ib~~l~~ibytes of space in the future. I'm genuinely interested, what are these little places?

**Comment by kevingranade** [comment]
If you're trying to set a precedent for more extensive use for git submodules, stop now, that's not happening.  
I was interested in seeing the outcome of using submodules for an optional dependency, but as I'm positively allergic to taking dependencies in the first place, I don't anticipate there being many opportunities for using them in the future.

I'm not clear what the mechanism would be for having people not download the entire project history. We can possibly shift out the tilesets, but even that would be a massive undertaking of debasing the entire project history to remove them from the main repository, and nothing else is coming to mind that is optional for download.

**Comment by jbytheway** [comment]
Since you asked, here are some of the annoyances from submodules I'm aware of:
* New users/developers need to do the extra commands to get everything.  This is yet another obstacle to the already-large issues with learning git that new contributors have, especially on a project like this where many contributors are not previously git users.
* If you clone the repo without the submodule and try to work somewhere without Internet access then it's too late to download the submodule.
* GitHub zip downloads of repos [don't include submodules](https://github.com/dear-github/dear-github/issues/214).
* `git archive` doesn't include submodules (which is the reason for the above point).
* Every system that automatically obtains and builds the code needs to be updated to take account of this.  This is not just every CI system we set up for the project, but also each one of the [many Linux distros](https://repology.org/project/cataclysm-dda/versions) that ship a CDDA package.
* If the module you reference ceases to exist or rewrites its history then historical versions of CDDA become impossible to build, because the referenced commits cannot be found.  Even if the referenced repo merely moves to a new host, then you must use little-known git configuration options to redirect to the new location, and you have to change that configuration depending on the point in history you're trying to build, which is messy at best.

**Comment by int-ua** [comment]
> stop now, that's not happening.

ok, I'll just copy the files.

> optional for download

It's becoming off-topic and irrelevant here but just to explain my opinion, in the order of being suitable for offloading into submodules (more to less):

* `data/mods/*`: as a person who mostly plays default/vanilla
* `android` and maybe other platform-specific folders 
* `lang/po`: as a person who plays without l10n. Although I've tried Transifex for my lang but it's just too slow. This folder is larger than everything else combined.

**Comment by int-ua** [comment]
Is there a benefit in all this convoluted mess that is cmake?

**Review by olanti-p — COMMENTED** [review]
Here's a bunch of changes that should make it compile

**Comment by int-ua** [comment]
Thank you! Let's see if I missed anything...

**Comment by int-ua** [comment]
fatal error: SDL_version.h: No such file or directory

Is it really a problem that was introduced by this PR? Is g++-7 build different in that it needs something like one of the configs in `CMakeModules/`?

File: CMakeLists.txt [hunk 1]
```diff
@@ -369,6 +369,8 @@ IF(USE_HOME_DIR)
 ENDIF(USE_HOME_DIR)
 
 add_subdirectory(src)
+add_executable(main src/main.cpp)
```
Comment (jbytheway) [inline CMakeLists.txt:372]: You don't want to add a new executable like this; it needs to be made a dependency of the existing executables.

zlib is already a dependency of the binaries in TILES mode, so you can see how this is done (in `src/CMakeLists.txt`, not this file) and mimic that.  Will need something like
```cmake
target_include_directories(cataclysm-common PUBLIC ${ZLIB_INCLUDE_DIRS})
target_link_libraries(cataclysm-common ${ZLIB_LIBRARIES})
```

Comment (int-ua) [reply CMakeLists.txt]: Not realizing there is a `src/CMakeLists.txt` was my major problem until now, thank you!

**Comment by int-ua** [comment]
Still `ld: ../src/CMakeFiles/cataclysm-tiles-common.dir/gzstream.cpp.o: undefined reference to symbol 'gzread'`

**Comment by int-ua** [comment]
`//lib/x86_64-linux-gnu/libz.so.1: error adding symbols: DSO missing from command line`
Ok, looks like `g++-7` is still missing `-lz` somewhere.

**Comment by int-ua** [comment]
`/usr/local/android-sdk-25.2.3/ndk-bundle/build/core/build-binary.mk:651: Android NDK: Module main depends on undefined modules: z`

**Comment by jbytheway** [comment]
I looked at the `src/CMakeLists.txt` a little more closely.  It appears that zlib is only being made a dependency of `cataclysm-tiles-common` when `DYNAMIC_LINKING` is `OFF` (which is not the default).  That dependency needs to be made unconditional; you can move the ZLIB parts down to the `target_include_directories` and `target_link_libraries` after `ENDIF (NOT DYNAMIC_LINKING)`.

**Pull request closed without merging**